cmd/root.go

```diff
@@ -433,40 +434,54 @@ func (te *cmdObjExpander) Expand() ([]*unstructured.Unstructured, error) {
 	}
 
 	_, vendorPath := manager.LibPaths()
-	libPath, mainPath, paramsPath, err := manager.EnvPaths(te.config.env)
+	libPath, mainPath, _, err := manager.EnvPaths(te.config.env)
 	if err != nil {
 		return nil, err
 	}
 
 	expander.FlagJpath = append([]string{string(vendorPath), string(libPath)}, expander.FlagJpath...)
 
-	componentPaths, err := manager.ComponentPaths()
+	namespacedComponentPaths, err := component.MakePathsByNamespace(te.config.fs, manager, te.config.cwd, te.config.env)
 	if err != nil {
 		return nil, errors.Wrap(err, "component paths")
 	}
 
-	baseObj, err := constructBaseObj(componentPaths, te.config.components)
-	if err != nil {
-		return nil, errors.Wrap(err, "construct base object")
-	}
-
 	//
 	// Set up ExtCodes to resolve runtime variables such as the environment namespace.
 	//
 
-	params := importParams(string(paramsPath))
 	envSpec, err := importEnv(manager, te.config.env)
 	if err != nil {
 		return nil, err
 	}
 
-	expander.ExtCodes = append([]string{baseObj, params, envSpec}, expander.ExtCodes...)
+	baseCodes := expander.ExtCodes
 
-	//
-	// Expand the ksonnet app as rendered for environment `env`.
-	//
+	slUnstructured := make([]*unstructured.Unstructured, 0)
+	for ns, componentPaths := range namespacedComponentPaths {
```

Comment:
The library paths are being repeatedly added because of this loop.

_i.e._ on _apply_, for every component namespace, we will be getting same sets of these messages:
```
DEBUG Adding jsonnet search path /Users/jyuen/workspace/go/src/github.com/ksonnet/ksonnet/foo/vendor
DEBUG Adding jsonnet search path /Users/jyuen/workspace/go/src/github.com/ksonnet/ksonnet/foo/lib/v1.7.0
```

Can you change this so that the lib paths are added only once?

Reply:
To clear this up, they are only added once for each invocation. The issue is that the `jsonnetVM` code has to run for each namespace separately. There way to solve it is to have the expander worker with multiple component namespaces and that might be more trouble than it is worth to get this functionality started.

Reply:
For posterity, here is what is generated if you filter out stderr.

```yaml
---
apiVersion: v1
kind: Service
metadata:
  name: guiroot
spec:
  ports:
  - port: 80
    targetPort: 80
  selector:
    app: guiroot
  type: ClusterIP
---
apiVersion: apps/v1beta1
kind: Deployment
metadata:
  name: guiroot
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: guiroot
    spec:
      containers:
      - image: gcr.io/heptio-images/ks-guestbook-demo:0.1
        name: guiroot
        ports:
        - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: gui
spec:
  ports:
  - port: 80
    targetPort: 80
  selector:
    app: gui
  type: ClusterIP
---
apiVersion: apps/v1beta1
kind: Deployment
metadata:
  name: gui
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: gui
    spec:
      containers:
      - image: gcr.io/heptio-images/ks-guestbook-demo:0.1
        name: gui
        ports:
        - containerPort: 80
```